src/sinks/util/sink.rs

```diff
@@ -105,141 +109,132 @@ where
+                return Poll::Ready(Ok(()));
+            }
 
+            // Try send batch.
+            let should_send = {
+                if (self.closing && !self.batch.is_empty()) || self.batch.was_full() {
+                    true
+                } else {
+                    self.linger
+                        .as_mut()
+                        .map(|linger| matches!(linger.poll_unpin(cx), Poll::Ready(())))
+                        .unwrap_or(false)
+                }
+            };
+            if should_send {
+                let service_ready = match self.service.poll_ready(cx) {
+                    Poll::Ready(Ok(())) => true,
+                    Poll::Ready(Err(error)) => return Poll::Ready(Err(error)),
+                    Poll::Pending => false,
+                };
+                if service_ready {
                     trace!("Service ready; Sending batch.");
+
                     let batch = self.batch.fresh_replace();
+                    self.linger = None;
 
                     let batch_size = batch.num_items();
                     let request = batch.finish();
+                    tokio::spawn(self.service.call(request, batch_size));
 
-                    let fut = self.service.call(request, batch_size).compat();
-                    tokio::spawn(fut);
+                    continue;
+                }
+            }
 
-                    // Remove the now-sent batch's linger timeout
-                    self.linger = None;
-                } else {
-                    // We have data but not a full batch and the linger time has not elapsed, so do
-                    // not sent a request yet. Instead, poll the inner service to drive progress
-                    // and defer readiness to the linger timeout if present.
-                    if let Some(linger) = &mut self.linger {
-                        self.service.poll_complete()?;
-
-                        // It's unlikely that we get past `try_ready` here because there's no way
-                        // for `should_send` to return false without polling a present `linger`. If
-                        // we do get past it, that means the timer expired within the tiny window
-                        // between the two `poll` calls and we should loop again because it's
-                        // probably time to send a batch.
-                        trace!("Polling batch linger.");
-                        try_ready!(linger.poll());
-                    } else {
-                        try_ready!(self.service.poll_complete());
-                    }
+            // Try move buffer to batch.
+            if self.buffer.is_some() && self.batch.is_empty() {
+                let item = self.buffer.take().unwrap();
+                self.as_mut().start_send(item)?;
+
+                if self.buffer.is_some() {
+                    unreachable!("Empty buffer overflowed.");
                 }
+
+                continue;
             }
```

Comment:
Does this fail the borrow checker (because `self` stays borrowed by `item`)?
```suggestion
            if self.batch.is_empty {
                if let Some(item) = self.buffer.take() {
                    self.as_mut().start_send(item)?;

                    if self.buffer.is_some() {
                        unreachable!("Empty buffer overflowed.");
                    }

                    continue;
                }
            }
```

Reply:
No, there is no borrow checker issues, I just do not like inner if-conditions :slightly_smiling_face: It would be nice to have something like `if let Some(item) = self.buffer.take() && self.batch.is_empty()`.
`unwrap` looks worse than inner if-conditions, I'll change.